include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h

```diff
@@ -64,22 +65,118 @@ struct PortInfo {
 /// Verification hook for verifying module like operations.
 LogicalResult verifyModuleLikeOpInterface(FModuleLike module);
 
-class InnerSymbolTable {
+namespace detail {
+LogicalResult verifyInnerRefs(Operation *op);
+} // namespace detail
 
+/// Table of inner_sym's
+class InnerSymbolTable {
 public:
   /// Return the name of the attribute used for inner symbol names.
   static StringRef getInnerSymbolAttrName() { return "inner_sym"; }
+
+  /// Build an inner_sym table for the given operation.
+  explicit InnerSymbolTable(Operation *op);
+
+  /// Look up a symbol with the specified name, returning null if no such
+  /// name exists. Names never include the @ on them.
+  Operation *lookup(StringRef name) const;
+  template <typename T>
+  T lookup(StringRef name) const {
+    return dyn_cast_or_null<T>(lookup(name));
+  }
+
+  /// Look up a symbol with the specified name, returning null if no such
+  /// name exists. Names never include the @ on them.
+  Operation *lookup(StringAttr name) const;
+  template <typename T>
+  T lookup(StringAttr name) const {
+    return dyn_cast_or_null<T>(lookup(name));
+  }
+
+private:
+  /// This is the operation this table is constructed for.
+  /// Must have InnerSymbolTable trait.
+  Operation *innerSymTblOp;
+
+  /// This maps names to operations with that inner symbol.
+  DenseMap<StringAttr, Operation *> symbolTable;
```

Comment:
Can we have the map with something like `DenseMap<SymbolNameAttr, Operation*> symbolTable`, such that there is no assumption that the inner sym is always a string. 
This will make it easier to transition to a new inner sym attribute to specify symbols per field etc.
https://github.com/llvm/circt/pull/3298/files

Reply:
Happy to change the key of the map, maybe this should just be `Attribute`? But I'll make it `SymbolNameAttr` for now, thanks (and thanks for looking out how these will interact!).

It's the value type that I'm unsure about how to best handle (presently, and with your change): it'd be good to have inner_sym's pointing to more than `Operation *`-- something like `AnnoTarget` although that's not quite right as these technically aren't only for annotations.  Just mostly.  Was planning to extend this to support resolving/tracking ports, and will want to handle per-field resolution as well.

Reply:
That's a good point. `AnnoTarget` seems like general enough. Maybe we can extend 
`AnnoTargetImpl::op` and `AnnoTargetImpl::portNo` for subfields also, the `portNo` can be reused/abused for `fieldID` .

Reply:
But anyway, these are all future extensions, we can keep them out of this PR!